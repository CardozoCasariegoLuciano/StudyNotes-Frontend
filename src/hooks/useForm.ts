import { ChangeEvent, useState } from "react";
import { FormValidator } from "../helpers/Validators/FormValidators";
import { IFormValidateData, IFormState } from "../interfaces/Forms.interface";
import { IFormValidationsErrors } from "../interfaces/Validator.interface";

const initialErrors = (form: IFormState) => {
  const result: IFormValidationsErrors = {};

  const formKeys = Object.keys(form);
  formKeys.forEach((name) => {
    result[name] = {};
  });

  return result;
};

export const useForm = <T extends IFormState>(
  initialForm: T,
  validations?: IFormValidateData,
  customMessagges?: IFormValidationsErrors
) => {
  const [formState, setFormState] = useState<T>(initialForm);
  const [formErrors, setFormErrors] = useState<IFormValidationsErrors>(
    initialErrors(formState)
  );
  const [_, setForceChange] = useState(false);

  const onInputChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const { name, value } = event.target;
    setFormState({
      ...formState,
      [name]: value,
    });

    if (value.length <= 0) {
      setFormErrors((actual) => ({ ...actual, [name]: {} }));
    } else {
      const optional = validations && formState[validations![name]?.equalTo!];
      setFormErrors(
        FormValidator(
          formErrors,
          name,
          value,
          validations!,
          optional,
          customMessagges
        )
      );
    }
  };

  const onBlur = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const { name, value } = event.target;
    const optional = validations && formState[validations[name]?.equalTo!];
    onInputChange(event);
    setFormErrors(
      FormValidator(
        formErrors,
        name,
        value,
        validations!,
        optional,
        customMessagges
      )
    );
  };

  const hasErrors = (): boolean => {
    let result = false;
    const formErrorsNames = Object.keys(formErrors);

    formErrorsNames.forEach((name) => {
      const value = Object.keys(formErrors[name]);
      if (value.length > 0) result = true;
    });

    return result;
  };

  const _findFormErrors = () => {
    const formStateNames = Object.keys(formState);

    formStateNames.forEach((name) => {
      {/* TODO: Ver como manejar eso del equalTo, esta raro asi dom 09 abr 2023 12:54:52  */}
      const optional = validations && formState[validations![name]?.equalTo!];
      const value = formState[name];
      setFormErrors(
        FormValidator(
          formErrors,
          name,
          value,
          validations!,
          optional,
          customMessagges
        )
      );
    });
  };

  const checkFormErrors = (ev: ChangeEvent<HTMLFormElement>) => {
    ev.preventDefault();
    setForceChange((actual) => !actual);
    _findFormErrors();
  };

  const onResetForm = () => {
    setFormState(initialForm);
  };

  const onCleanInput = (name: string) => {
    setFormState((actual) => ({ ...actual, [name]: "" }));
    setFormErrors((actual) => ({ ...actual, [name]: {} }));
  };

  return {
    checkFormErrors,
    ...formState,
    formState,
    formErrors,
    hasErrors,
    onInputChange,
    onBlur,
    onResetForm,
    onCleanInput,
  };
};
