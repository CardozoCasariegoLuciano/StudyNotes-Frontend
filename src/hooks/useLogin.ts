import axios, { AxiosError } from "axios";
import { UrlBank } from "../helpers/URLs";
import {
  APISuccessAuth,
  LoginFormData,
  LoginResult,
} from "../interfaces/API_auth.interface";
import { APIError } from "../interfaces/API_response.interface";

const useLogin = ():LoginResult => {
  let apiError: APIError | null = null;

  const login = async (loginData: LoginFormData): Promise<APIError | null> => {
    try {
      const resp = await axios.post(UrlBank.auth.login, loginData);
      const data = resp.data as APISuccessAuth;

      localStorage.setItem("token", JSON.stringify(data.data.token));

      apiError = null;
    } catch (err) {
      const axiosError = err as AxiosError;
      apiError = axiosError.response?.data as APIError;
    }
    return apiError;
  };

  return { login };
};

export default useLogin;
