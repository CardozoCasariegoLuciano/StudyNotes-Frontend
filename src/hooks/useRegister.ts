import axios, { AxiosError } from "axios";
import { UrlBank } from "../helpers/URLs";
import {
  APISuccessAuth,
  RegisterFormData,
  RegisterResult,
} from "../interfaces/API_auth.interface";
import { APIError } from "../interfaces/API_response.interface";

const useRegister = (): RegisterResult => {
  let apiError: APIError | null = null;

  const register = async (
    formData: RegisterFormData
  ): Promise<APIError | null> => {
    try {
      const registerData = {
        name: formData.name,
        email: formData.email,
        password: formData.password,
        confirmPassword: formData.confirmation,
      };

      const resp = await axios.post(UrlBank.auth.register, registerData);
      const data = resp.data as APISuccessAuth;

      localStorage.setItem("token", JSON.stringify(data.data.token));
      localStorage.setItem(
        "user",
        JSON.stringify({ name: formData.name, email: formData.email })
      );

      apiError = null;
    } catch (err) {
      const axiosError = err as AxiosError;
      apiError = axiosError.response?.data as APIError;
    }
    return apiError;
  };

  return { register };
};

export default useRegister;
