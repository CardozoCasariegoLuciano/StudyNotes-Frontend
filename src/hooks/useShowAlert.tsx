import { useRef, useState } from "react";

type Ialert = {
  isVisible: boolean;
  text: string;
};

const alertStyles = {
  container: {
    border: "1px solid red",
    backgroundColor: "tomato",
    padding: "5px 20px",
    borderRadius: "8px",
    display: "flex",
    justifyContent: "space-between",
    gap: "15px",
    color: "white",
  },
  X: {
    cursor: "pointer",
  },
};

export const useShowAlert = (time: number = 5500) => {
  const timeoutRef = useRef<number | undefined>();
  const [modal, setModal] = useState<Ialert>({
    isVisible: false,
    text: "",
  });

  const cleanAlert = () => {
    setModal({ isVisible: false, text: "" });
  };

  const showModal = (value: string): void => {
    timeoutRef.current && clearInterval(timeoutRef.current);
    setModal({ isVisible: true, text: value });
    timeoutRef.current = window.setTimeout(() => {
      cleanAlert();
    }, time);
  };

  const showAlert = () => {
    const alert = (
      <div style={alertStyles.container}>
        <p>{modal.text} </p>
        <span
          style={alertStyles.X}
          onClick={() => {
            timeoutRef.current && clearInterval(timeoutRef.current);
            cleanAlert();
          }}
        >
          X
        </span>
      </div>
    );
    return modal.isVisible && alert;
  };

  return {
    showModal,
    alertText: modal.text,
    isVisible: modal.isVisible,
    showAlert,
  } as const;
};
