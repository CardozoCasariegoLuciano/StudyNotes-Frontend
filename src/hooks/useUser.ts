import { useContext } from "react";
import { UserContext } from "../contexts/userContext";
import { UserProviderData } from "../interfaces/userContext.interface";

const useUser = () => {
  const contextValue = useContext(UserContext) as UserProviderData;
  if (contextValue == null) {
    throw new Error("No se puede acceder al contexto");
  }

  return contextValue;
};

export default useUser;
