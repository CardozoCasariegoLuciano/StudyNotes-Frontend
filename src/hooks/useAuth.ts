import axios, { AxiosError } from "axios";
import { useNavigate } from "react-router-dom";
import { manageAPIErrors } from "../helpers/apiErrors";
import { UrlBank } from "../helpers/URLs";
import { removeStorage, setStorage } from "../helpers/webStorage";
import {
  APISuccessLogin,
  APISuccessRegister,
  AuthReuslt,
  LoginFormData,
  RegisterFormData,
} from "../interfaces/API_auth.interface";
import { APIError } from "../interfaces/API_response.interface";
import useUser from "./useUser";

const useAuth = (): AuthReuslt => {
  const { logOut: logout} = useUser();
  const navigate = useNavigate();
  let apiError: APIError | null = null;

  const login = async (loginData: LoginFormData): Promise<APIError | null> => {
    const reqBody = {
      email: loginData.email,
      password: loginData.password,
    };

    try {
      const axiosResp = await axios.post(UrlBank.auth.login, reqBody);
      const res = axiosResp.data as APISuccessLogin;

      setStorage("token", JSON.stringify(res.data.token), loginData.remember);

      apiError = null;
    } catch (err) {
      apiError = manageAPIErrors(err as AxiosError);
    }
    return apiError;
  };

  const register = async (
    formData: RegisterFormData
  ): Promise<APIError | null> => {
    try {
      const registerData = {
        name: formData.name,
        email: formData.email,
        password: formData.password,
        confirmPassword: formData.confirmation,
      };

      const axiosResponse = await axios.post(
        UrlBank.auth.register,
        registerData
      );
      const resp = axiosResponse?.data as APISuccessRegister;

      setStorage("token", JSON.stringify(resp.data.token), true);

      apiError = null;
    } catch (err) {
      apiError = manageAPIErrors(err as AxiosError);
    }
    return apiError;
  };

  const logOut = () => {
    removeStorage("token");
    logout();
    navigate("/auth/login");
  };

  return { login, register, logOut };
};

export default useAuth;
