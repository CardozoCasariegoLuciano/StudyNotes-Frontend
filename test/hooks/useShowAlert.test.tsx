import { render, renderHook, screen, waitFor } from '@testing-library/react';
import {ReactElement} from 'react';
import { act } from 'react-dom/test-utils';
import { useShowAlert } from '../../src/hooks/useShowAlert';

describe('useShowAlert tests cases', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  test('Should show the initial values', () => {
    const { result } = renderHook(() => useShowAlert());

    expect(result.current.alertText).toBe('');
    expect(result.current.isVisible).toBeFalsy();
  });

  test('alertText and isVisible mush change durin initial time', () => {
    const time = 500;
    const text = 'TEST';

    const { result } = renderHook(() => useShowAlert(time));

    act(() => {
      result.current.showModal(text);
    });

    expect(result.current.alertText).toBe(text);
    expect(result.current.isVisible).toBeTruthy();

    act(() => {
      jest.advanceTimersByTime(time);
    });

    expect(result.current.alertText).toBe('');
    expect(result.current.isVisible).toBeFalsy();
  });

  test('ShowAlert mush display the alert component', async () => {
    const time = 500;
    const text = 'TExT';

    const { result } = renderHook(() => useShowAlert(time));

    act(() => {
      result.current.showModal(text);
    });

    const temp = result.current.showAlert();
    render(temp as ReactElement);

    expect(temp).toMatchSnapshot();
  });
});
