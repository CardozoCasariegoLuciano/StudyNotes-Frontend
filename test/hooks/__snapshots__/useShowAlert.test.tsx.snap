// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useShowAlert tests cases ShowAlert mush display the alert component 1`] = `
<div
  style={
    {
      "backgroundColor": "tomato",
      "border": "1px solid red",
      "borderRadius": "8px",
      "color": "white",
      "display": "flex",
      "gap": "15px",
      "justifyContent": "space-between",
      "padding": "5px 20px",
    }
  }
>
  <p>
    TExT
     
  </p>
  <span
    onClick={[Function]}
    style={
      {
        "cursor": "pointer",
      }
    }
  >
    X
  </span>
</div>
`;
